openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,435 @@
+
+	// we do not want to change guaranteed pods resource allocation, because it should be managed by
+	// relevant managers(CPU and memory) under the kubelet
+	if podQoSClass == coreapi.PodQOSGuaranteed {
+		pod.Annotations[workloadAdmissionWarning] = admissionWarningGuaranteedPod
+		return nil
+	}
+
+	// before we update the pod available under admission attributes, we need to verify that deletion of the CPU request
+	// will not change the pod QoS class, otherwise skip pod mutation
+	// 1. Copy the pod
+	// 2. Delete CPUs requests for all containers under the pod
+	// 3. Get modified pod QoS class
+	// 4. Verify that the pod QoS class before and after the modification stay the same
+	// 5. Update the pod under admission attributes
+	podCopy := pod.DeepCopy()
+	updatePodResources(podCopy, workloadType, podQoSClass)
+
+	allContainersCopy := append([]coreapi.Container{}, podCopy.Spec.InitContainers...)
+	allContainersCopy = append(allContainersCopy, podCopy.Spec.Containers...)
+	podQoSClassAfterModification := getPodQoSClass(allContainersCopy)
+
+	if podQoSClass != podQoSClassAfterModification {
+		pod.Annotations[workloadAdmissionWarning] = fmt.Sprintf(admissionWarningChangePodQoSClass, podQoSClass, podQoSClassAfterModification)
+		return nil
+	}
+
+	updatePodResources(pod, workloadType, podQoSClass)
+
+	return nil
+}
+
+func (a *managementCPUsOverride) waitForSyncedStore(timeout <-chan time.Time) bool {
+	for !a.nsListerSynced() || !a.nodeListSynced() {
+		select {
+		case <-time.After(100 * time.Millisecond):
+		case <-timeout:
+			return a.nsListerSynced() && a.nodeListSynced()
+		}
+	}
+
+	return true
+}
+
+func updatePodResources(pod *coreapi.Pod, workloadType string, class coreapi.PodQOSClass) {
+	if pod.Annotations == nil {
+		pod.Annotations = map[string]string{}
+	}
+
+	// update init containers resources
+	updateContainersResources(pod.Spec.InitContainers, pod.Annotations, workloadType, class)
+
+	// update app containers resources
+	updateContainersResources(pod.Spec.Containers, pod.Annotations, workloadType, class)
+
+	// set workload annotation
+	workloadAnnotation := fmt.Sprintf("%s%s", podWorkloadAnnotationPrefix, workloadType)
+	pod.Annotations[workloadAnnotation] = fmt.Sprintf(`{"%s":"%s"}`, podWorkloadAnnotationEffect, workloadEffectPreferredDuringScheduling)
+}
+
+func updateContainersResources(containers []coreapi.Container, podAnnotations map[string]string, workloadType string, podQoSClass coreapi.PodQOSClass) {
+	for i := range containers {
+		c := &containers[i]
+		cpusAnnotationKey := fmt.Sprintf("%s/%s", containerCPUSharesAnnotation, c.Name)
+
+		if podQoSClass == coreapi.PodQOSBestEffort {
+			podAnnotations[cpusAnnotationKey] = "2"
+			continue
+		}
+
+		if c.Resources.Requests != nil {
+			// if the kubelet deleted container CPUs requests from resources for static pods,
+			// the admission plugin will not do additional updates
+			if _, ok := c.Resources.Requests[coreapi.ResourceCPU]; !ok {
```

Comment:
making this interact with another component (kubelet) is bad (non-local logic). Why not just ignore mirror pods in the main func?

Reply:
Can we remove the comment now that static pods are ignored?

Reply:
done